flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java

```diff
@@ -257,7 +289,32 @@ public void onEventTime(InternalTimer<KEY, VoidNamespace> timer) throws Exceptio
 
 	@Override
 	public void onProcessingTime(InternalTimer<KEY, VoidNamespace> timer) throws Exception {
-		// not used
+		NFA<IN> nfa = getNFA();
+
+		// emit the events in order
+		for (IN event : sort(bufferedEvents.get())) {
+			processEvent(nfa, event, getProcessingTimeService().getCurrentProcessingTime());
+		}
+
+		// remove all buffered rows
+		bufferedEvents.clear();
+
+		updateNFA(nfa);
+	}
+
+	private Iterable<IN> sort(Iterable<IN> iter) {
+		if (comparator == null) {
+			return iter;
+		} else {
+			// insert all events into the sort buffer
+			List<IN> sortBuffer = new ArrayList<>();
```

Comment:
As we dropped support of Java 7, we could use `Stream` here. That way we won't need to create the temporary buffer.

Sth like:

	private Stream<IN> sort(Iterable<IN> iter) {
		Stream<IN> stream = StreamSupport.stream(iter.spliterator(), false);
		if (comparator == null) {
			return stream;
		} else {
			return stream.sorted(comparator);
    }

And use it like this:

	sort(bufferedEvents.get()).forEachOrdered(
		event -> processEvent(nfa, event, getProcessingTimeService().getCurrentProcessingTime())
	);

What do you think?